### PR TITLE
secp384r1 Solinas' Implementation and Power asm acceleration 

### PR DESCRIPTION
**Pull request by rohanmclure**
Adopt a 56-bit redundant-limb implementation, utilising Solinas' reduction for the NIST p384 curve. In keeping with the ecp_nistp224.c and ecp_nistp521.c implementations, allow for future processor-specific acceleration of  felem_{square,mul} methods by means of a simple ABI.

This implementation yielded performance gains of 446% and 106% on digital signing and verification respectively (tested on the Apple M1 Max), and substantial gains on all other platforms tested.

Also provide an asm implementation for Power processors P9 and above, yielding an additional speedup of 9.5% on signature generation.

### CODE REVIEW

**Comment by rohanmclure** [comment]
This PR updates the documentation on `EC_GFp_nistp384_method()`, as these functions should all be deprecated. Doing so would appear to be to introduce an anachronism. If anyone has any suggestions on how to avoid this I'd appreciate them.

File: include/openssl/ec.h [hunk 1]
```diff
@@ -144,6 +144,11 @@ OSSL_DEPRECATEDIN_3_0 const EC_METHOD *EC_GFp_nistp224_method(void);
  */
 OSSL_DEPRECATEDIN_3_0 const EC_METHOD *EC_GFp_nistp256_method(void);
 
+/** Returns 64-bit optimized methods for nistp384
+ *  \return  EC_METHOD object
+ */
+OSSL_DEPRECATEDIN_3_0 const EC_METHOD *EC_GFp_nistp384_method(void);
+
```
Comment (t8m) [inline include/openssl/ec.h:151]: Just avoid adding the public function and create an internal non-deprecated function. IMO there is no need to have this function public.

**Review by t8m — CHANGES_REQUESTED** [review]

File: include/openssl/ec.h (hunk 1 above)
Comment (rohanmclure) [reply include/openssl/ec.h]: Done! Thanks

File: crypto/ec/ec_local.h [hunk 2]
```diff
@@ -551,6 +555,27 @@ int ossl_ec_GFp_nistp256_points_mul(const EC_GROUP *group, EC_POINT *r,
...
+                             const BIGNUM *scalar, size_t num,
+                             const EC_POINT *points[], const BIGNUM *scalars[],
+                             BN_CTX *);
+int ossl_ec_GFp_nistp384_points_mul(const EC_GROUP *group, EC_POINT *r,
+                                    const BIGNUM *scalar, size_t num,
+                                    const EC_POINT *points[],
+                                    const BIGNUM *scalars[], BN_CTX *ctx);
+int ossl_ec_GFp_nistp384_precompute_mult(EC_GROUP *group, BN_CTX *ctx);
+int ossl_ec_GFp_nistp384_have_precompute_mult(const EC_GROUP *group);
+const EC_METHOD *EC_GFp_nistp384_method(void);
```
Comment (t8m) [inline crypto/ec/ec_local.h:577]: All the newly added internal non-static functions should be with lower case ossl_<module> prefix - i.e. `ossl_ec_GFp_nistp384_method` and similarly also for those two other functions added above.

If you want, you can also rename the other internal EC_ functions in a separate commit, to make the function name style uniform. But that is not required for this PR.

File: crypto/ec/ec_lib.c [hunk 3]
```diff
@@ -99,12 +99,16 @@ void EC_pre_comp_free(EC_GROUP *group)
...
+    case PCT_nistp384:
+        EC_nistp384_pre_comp_free(group->pre_comp.nistp384);
+        break;
     case PCT_nistp521:
         EC_nistp521_pre_comp_free(group->pre_comp.nistp521);
         break;
 #else
     case PCT_nistp224:
     case PCT_nistp256:
+	case PCT_nistp384:
```
Comment (beldmit) [inline crypto/ec/ec_lib.c:111]: Nit: alignment

File: crypto/ec/ec_local.h [hunk 4]
```diff
@@ -264,12 +265,13 @@ struct ec_group_st {
...
     enum {
         PCT_none,
-        PCT_nistp224, PCT_nistp256, PCT_nistp521, PCT_nistz256,
+        PCT_nistp224, PCT_nistp256, PCT_nistp384, PCT_nistp521, PCT_nistz256,
         PCT_ec
     } pre_comp_type;
     union {
         NISTP224_PRE_COMP *nistp224;
         NISTP256_PRE_COMP *nistp256;
+		NISTP384_PRE_COMP *nistp384;
```
Comment (beldmit) [inline crypto/ec/ec_local.h:274]: Nit: alignment

**Review by beldmit — CHANGES_REQUESTED** [review]

File: crypto/ec/ec_local.h (hunk 4 above)
Comment (rohanmclure) [reply crypto/ec/ec_local.h]: Corrected, thanks.

**Review by t8m — APPROVED** [review]

File: crypto/ec/ecp_nistp384.c [hunk 5]
```diff
@@ -0,0 +1,1935 @@
...
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/*
+ * ECDSA low level APIs are deprecated for public use, but still ok for
+ * internal use.
+ */
+#include "internal/deprecated.h"
```
Comment (paulidale) [inline crypto/ec/ecp_nistp384.c:14]: Why?  This is a new file, it should even mention deprecation.

File: crypto/ec/ecp_nistp384.c [hunk 6]
```diff
@@ -0,0 +1,1935 @@
...
+}
+
+/*-
+ * Field operations
+ * ----------------
+ */
+
+static void felem_one(felem out)
+{
+    unsigned i;
```
Comment (paulidale) [inline crypto/ec/ecp_nistp384.c:166]: Our coding style uses `unsigned int` here and elsewhere.

Also a blank line separating declarations and code, again here and elsewhere.

File: crypto/ec/ecp_nistp384.c [hunk 7]
```diff
@@ -0,0 +1,1935 @@
...
+    out[0] = 1;
+    for (i = 1; i < NLIMBS; i++)
+        out[i] = 0;
```
Comment (paulidale) [inline crypto/ec/ecp_nistp384.c:169]: Why not use `memset(3)` ?

File: crypto/ec/ecp_nistp384.c [hunk 8]
```diff
@@ -0,0 +1,1935 @@
...
+}
+
+static void felem_assign(felem out, const felem in)
+{
+    unsigned i;
+    for (i = 0; i < NLIMBS; i++)
+        out[i] = in[i];
```
Comment (paulidale) [inline crypto/ec/ecp_nistp384.c:176]: Why not use `memcpy(3)` ?

File: crypto/ec/ecp_nistp384.c [hunk 9]
```diff
@@ -0,0 +1,1935 @@
...
+/*-
+ * felem_neg sets |out| to |-in|
+ * On entry:
+ *   in[i] < 2^60 - 2^29
+ * On exit:
+ *   out[i] < 2^60
+ */
+static void felem_neg(felem out, const felem in)
+{
+    /* In order to prevent underflow, we subtract from 0 mod p. */
```
Comment (paulidale) [inline crypto/ec/ecp_nistp384.c:220]: More explanation would be nice.  The subtraction doesn't look to be from zero.
_subtract from p which is 0 mod p_ perhaps?

File: crypto/ec/ecp_nistp384.c [hunk 10]
```diff
@@ -0,0 +1,1935 @@
...
+ * felem_diff64 subtracts |in| from |out|
+ * On entry:
+ *   in[i] < 2^60 - 2^52 - 2^4
+ * On exit:
+ *   out[i] < out[i] + 2^60 + 2^44
+ */
+static void felem_diff64(felem out, const felem in)
+{
+    /*
+     * In order to prevent underflow, we add 0 mod p before subtracting.
```
Comment (paulidale) [inline crypto/ec/ecp_nistp384.c:252]: Again, the comment could be improved.

File: crypto/ec/ecp_nistp384.c [hunk 11]
```diff
@@ -0,0 +1,1935 @@
...
+# define felem_mul felem_mul_p
+#else
+# define felem_square felem_square_ref
+# define felem_mul felem_mul_ref
+#endif
+
+static ossl_inline void felem_square_reduce(felem out, const felem in)
+{
+    widefelem tmp;
+    felem_square(tmp, in);
```
Comment (paulidale) [inline crypto/ec/ecp_nistp384.c:676]: blank line above -- blank line to separate declarations and code.

File: crypto/ec/ecp_nistp384.c [hunk 12]
```diff
@@ -0,0 +1,1935 @@
...
+    for (i = 0; i < 94; i++)
+        felem_square_reduce(ftmp, ftmp); /* 2^381 + ... + 2^127 + 2^125 + ... + 2^94 */
+    felem_mul_reduce(ftmp, ftmp6, ftmp); /* 2^381 + ... + 2^127 + 2^125 + ... + 2^94 + 2^29 + ... + 2^0 */
+
+    for (i = 0; i < 2; i++)
+        felem_square_reduce(ftmp, ftmp); /* 2^383 + ... + 2^129 + 2^127 + ... + 2^96 + 2^31 + ... + 2^2 */
+    felem_mul_reduce(ftmp, in, ftmp);    /* 2^383 + ... + 2^129 + 2^127 + ... + 2^96 + 2^31 + ... + 2^2 + 2^0 */
+
+    for (i = 0; i < NLIMBS; i++)
+        out[i] = ftmp[i];
```
Comment (paulidale) [inline crypto/ec/ecp_nistp384.c:762]: `memcpy(3)`?

File: crypto/ec/ecp_nistp384.c [hunk 13]
```diff
@@ -0,0 +1,1935 @@
...
+ * and 0 otherwise. We know that field elements are reduced to
+ * 0 < in < 2p, so we only need to check two cases:
+ * 0 and 2^384 - 2^128 - 2^96 + 2^32 - 1
+ *   in[k] < 2^56, k < 6
+ *   in[6] <= 2^48
+ */
+static limb felem_is_zero(const felem in)
+{
+    limb zero, p384;
+    zero = in[0] | in[1] | in[2] | in[3] | in[4] | in[5] | in[6];
```
Comment (paulidale) [inline crypto/ec/ecp_nistp384.c:776]: blank line above.

File: crypto/ec/ecp_nistp384.c [hunk 14]
```diff
@@ -0,0 +1,1935 @@
...
+ * Assume we've removed all redundant bits.
+ * On entry:
+ *   in[k] < 2^56, k < 6
+ *   in[6] <= 2^48
+ */
+static void felem_contract(felem out, const felem in)
+{
+    static const int64_t two56 = ((limb) 1) << 56;
+
+    /* We know for a fact that 0 <= in < 2*p, for p = 2^384 - 2^128 - 2^96 + 2^32 - 1
```
Comment (paulidale) [inline crypto/ec/ecp_nistp384.c:803]: Comment style nit:
```c
/*
 * We know for a fact that 0 <= in < 2*p, for p = 2^384 - 2^128 - 2^96 + 2^32 - 1
 * Perform two successive, idempotent subtractions to reduce if in >= p.
 */
```

File: crypto/ec/ecp_nistp384.c [hunk 15]
```diff
@@ -0,0 +1,1935 @@
...
+    tmp[3] &= 0x00ffffffffffffff;
+
+    tmp[5] += tmp[4] >> 56;
+    tmp[4] &= 0x00ffffffffffffff;
+
+    tmp[6] += tmp[5] >> 56;
+    tmp[5] &= 0x00ffffffffffffff;
+
+    for (i = 0; i < NLIMBS; i++)
+       out[i] = tmp[i];
```
Comment (paulidale) [inline crypto/ec/ecp_nistp384.c:923]: `memcpy(3)`?

File: crypto/ec/ecp_nistp384.c [hunk 16]
```diff
@@ -0,0 +1,1935 @@
...
+}
+
+/*-
+ * Group operations
+ * ----------------
+ *
+ * Building on top of the field operations we have the operations on the
+ * elliptic curve group itself. Points on the curve are represented in Jacobian
+ * coordinates */
```
Comment (paulidale) [inline crypto/ec/ecp_nistp384.c:932]: style nit: close the comment on a new line.

File: crypto/ec/ecp_nistp384.c [hunk 17]
```diff
@@ -0,0 +1,1935 @@
...
+
+/*-
+ * point_double calculates 2*(x_in, y_in, z_in)
+ *
+ * The method is taken from:
+ *   http://hyperelliptic.org/EFD/g1p/auto-shortw-jacobian-3.html#doubling-dbl-2001-b
+ *
+ * Outputs can equal corresponding inputs, i.e., x_out == x_in is allowed.
+ * while x_out == y_in is not (maybe this works, but it's not tested). */
```
Comment (paulidale) [inline crypto/ec/ecp_nistp384.c:941]: style nit: close the comment on a new line.

File: crypto/ec/ecp_nistp384.c [hunk 18]
```diff
@@ -0,0 +1,1935 @@
...
+}
+
+/* copy_conditional copies in to out iff mask is all ones. */
+static void copy_conditional(felem out, const felem in, limb mask)
+{
+    unsigned i;
+    for (i = 0; i < NLIMBS; i++) {
+        const limb tmp = mask & (in[i] ^ out[i]);
+        out[i] ^= tmp;
+    }
```
Comment (paulidale) [inline crypto/ec/ecp_nistp384.c:999]: Is it worthwhile using our constant time #include for this?
Not a firm request but worth thinking about.

File: crypto/ec/ecp_nistp384.c [hunk 19]
```diff
@@ -0,0 +1,1935 @@
...
+ *    11 | 1 0 1 1 | (2^285 + 2^95 + 1)G
+ *    12 | 1 1 0 0 | (2^285 + 2^190)G
+ *    13 | 1 1 0 1 | (2^285 + 2^190 + 1)G
+ *    14 | 1 1 1 0 | (2^285 + 2^190 + 2^95)G
+ *    15 | 1 1 1 1 | (2^285 + 2^190 + 2^95 + 1)G
+ *
+ * The reason for this is so that we can clock bits into four different
+ * locations when doing simple scalar multiplies against the base point.
+ *
+ * Tables for other points have table[i] = iG for i in 0 .. 16. */
```
Comment (paulidale) [inline crypto/ec/ecp_nistp384.c:1187]: nit: close comment on new line.

File: crypto/ec/ecp_nistp384.c [hunk 20]
```diff
@@ -0,0 +1,1935 @@
...
+{
+    unsigned i, j;
+    limb *outlimbs = &out[0][0];
+
+    memset(out, 0, sizeof(*out) * 3);
+
+    for (i = 0; i < size; i++) {
+        const limb *inlimbs = &pre_comp[i][0][0];
+        limb mask = i ^ idx;
+        mask |= mask >> 4;
```
Comment (paulidale) [inline crypto/ec/ecp_nistp384.c:1287]: blank line separating code from declarations.

File: crypto/ec/ecp_nistp384.c [hunk 21]
```diff
@@ -0,0 +1,1935 @@
...
+        mask |= mask >> 2;
+        mask |= mask >> 1;
+        mask &= 1;
+        mask--;
+        for (j = 0; j < NLIMBS * 3; j++)
+            outlimbs[j] |= inlimbs[j] & mask;
```
Comment (paulidale) [inline crypto/ec/ecp_nistp384.c:1293]: Again, would our constant time support be useful here?

File: crypto/ec/ecp_nistp384.c [hunk 22]
```diff
@@ -0,0 +1,1935 @@
...
+                                 * round */
+    for (i = (num_points ? 380 : 98); i >= 0; --i) {
+        /* double */
+        if (!skip)
+            point_double(nq[0], nq[1], nq[2], nq[0], nq[1], nq[2]);
+
+        /* add multiples of the generator */
+        if (gen_mul && (i <= 98)) {
+            bits = get_bit(g_scalar, i + 285) << 3;
+   if (i < 95) {
```
Comment (paulidale) [inline crypto/ec/ecp_nistp384.c:1342]: Indentation looks off here.

File: crypto/ec/ecp_nistp384.c [hunk 23]
```diff
@@ -0,0 +1,1935 @@
...
+          bits |= get_bit(g_scalar, i + 95) << 1;
+          bits |= get_bit(g_scalar, i);
+   }
+            /* select the point to add, in constant time */
+            select_point(bits, 16, g_pre_comp, tmp);
+            if (!skip) {
+                /* The 1 argument below is for "mixed" */
+                point_add(nq[0], nq[1], nq[2],
+                          nq[0], nq[1], nq[2], 1,
+        tmp[0], tmp[1], tmp[2]);
```
Comment (paulidale) [inline crypto/ec/ecp_nistp384.c:1353]: indentation??

File: crypto/ec/ecp_nistp384.c [hunk 24]
```diff
@@ -0,0 +1,1935 @@
...
+                 */
+                select_point(digit, 17, pre_comp[num], tmp);
+                felem_neg(tmp[3], tmp[1]); /* (X, -Y, Z) is the negative
+                                            * point */
+                copy_conditional(tmp[1], tmp[3], (-(limb) sign));
+
+                if (!skip) {
+                    point_add(nq[0], nq[1], nq[2],
+                              nq[0], nq[1], nq[2], mixed,
+         tmp[0], tmp[1], tmp[2]);
```
Comment (paulidale) [inline crypto/ec/ecp_nistp384.c:1383]: indentation

File: crypto/ec/ecp_nistp384.c [hunk 25]
```diff
@@ -0,0 +1,1935 @@
...
+    if (p == NULL)
+        return;
+
+    CRYPTO_DOWN_REF(&p->references, &i);
+    REF_PRINT_COUNT("ossl_ec_nistp384", p);
+    if (i > 0)
+        return;
+    REF_ASSERT_ISNT(i < 0);
+
+    CRYPTO_THREAD_lock_free(p->lock);
```
Comment (paulidale) [inline crypto/ec/ecp_nistp384.c:1507]: Need to free `p->references` using `CRYPTO_FREE_REF`.
On most platforms this is a no-op but not on all.

File: crypto/ec/ecp_nistp384.c [hunk 26]
```diff
@@ -0,0 +1,1935 @@
...
+
+/******************************************************************************/
+/*
+ * OPENSSL EC_METHOD FUNCTIONS
+ */
+
+int ossl_ec_GFp_nistp384_group_init(EC_GROUP *group)
+{
+    int ret;
+    ret = ossl_ec_GFp_simple_group_init(group);
```
Comment (paulidale) [inline crypto/ec/ecp_nistp384.c:1519]: blank line above.

File: crypto/ec/ecp_nistp384.c [hunk 27]
```diff
@@ -0,0 +1,1935 @@
...
+                    }
+                }
+            }
+        }
+        if (mixed)
+            make_points_affine(num_points * 17, pre_comp[0], tmp_felems);
+    }
+
+    /* the scalar for the generator */
+    if ((scalar != NULL) && (have_pre_comp)) {
```
Comment (paulidale) [inline crypto/ec/ecp_nistp384.c:1783]: We'd not normally have the inner parenthesis, here and elsewhere.

**Review by paulidale — CHANGES_REQUESTED** [review]

File: crypto/ec/ecp_nistp384.c (hunk 5 above)
Comment (t8m) [reply crypto/ec/ecp_nistp384.c:14]: Well it could use some api internally that is deprecated, however I do not see any such api in use here.

Comment (rohanmclure) [reply crypto/ec/ecp_nistp384.c:14]: Apologies, this is residual from v1, where I had deprecation annotations on ossl_ec_nistp384_method().

**Comment by rohanmclure** [comment]
Thanks @t8m , @beldmit , @paulidale for your comments thus far. I will get to them at the beginning of next week, but am on leave until then.

File: crypto/ec/ecp_nistp384.c (hunk 9 above)
Comment (rohanmclure) [reply crypto/ec/ecp_nistp384.c:220]: I meant subtract from the congruence class containing zero. Let's change that to 'subtract from a multiple of p'.

File: crypto/ec/ecp_nistp384.c (hunk 18 above)
Comment (rohanmclure) [reply crypto/ec/ecp_nistp384.c:999]: There aren't any real fit-for-purpose methods there. For example to use `constant_time_cond_swap_buff` we would need to copy the contents of `in` first. This method occurs in a hot path, so keeping original implementation.

File: crypto/ec/ecp_nistp384.c (hunk 21 above)
Comment (rohanmclure) [reply crypto/ec/ecp_nistp384.c:1293]: As `select_point` is used in a hot path, and so halves the performance of generating digital signatures unfortunately. Keeping original implementation.

File: crypto/ec/ecp_nistp384.c [hunk 28]
```diff
@@ -0,0 +1,1937 @@
...
+    output[5] += output[4] >> 56;   /*-
+                                     * output[5] < output_old[5] + 2^16 + 1
+                                     *           < 2^56 + 2^16 + 1
+                                     */
+    output[4] &= 0x00ffffffffffffff;
+
+    output[6] += output[5] >> 56;   /* output[6] < 2^48 + 1 */
+    output[5] &= 0x00ffffffffffffff;
+
+    for (i = 0; i < NLIMBS; i++)
```
Comment (paulidale) [inline crypto/ec/ecp_nistp384.c:611]: `memcpy(3)` ?

File: crypto/ec/ecp_nistp384.c [hunk 29]
```diff
@@ -0,0 +1,1937 @@
...
+
+    /*
+     * We know for a fact that 0 <= in < 2*p, for p = 2^384 - 2^128 - 2^96 + 2^32 - 1
+     * Perform two successive, idempotent subtractions to reduce if in >= p.
+     */
+
+    int64_t tmp[NLIMBS], a;
+    unsigned int i;
+
+    for (i = 0; i < NLIMBS; i++)
```
Comment (paulidale) [inline crypto/ec/ecp_nistp384.c:806]: `memcpyt(3)` ?

File: crypto/ec/ecp_nistp384.c [hunk 30]
```diff
@@ -0,0 +1,1937 @@
...
+    tmp[4] &= 0x00ffffffffffffff;
+
+    tmp[6] += tmp[5] >> 56; /* tmp[6] < 2^48 */
+    tmp[5] &= 0x00ffffffffffffff;
+
+    /*
+     * Case 2: a = all ones if p <= in < 2^384, 0 otherwise
+     */
+
+    int64_t cond[5];
```
Comment (paulidale) [inline crypto/ec/ecp_nistp384.c:845]: Please move this up to the beginning of the function, we don't mix code and declarations.

**Review by paulidale — CHANGES_REQUESTED** [review]

File: crypto/ec/ecp_nistp384.c (hunk 19 above)
Comment (paulidale) [reply crypto/ec/ecp_nistp384.c:1186]: This one was skipped.

File: crypto/ec/ecp_nistp384.c (hunk 28 above)
Comment (rohanmclure) [reply crypto/ec/ecp_nistp384.c:611]: Members in `output` and `out` have different sizes, so I'll keep the loop of assignments here.

File: crypto/ec/ecp_nistp384.c [hunk 31]
```diff
@@ -0,0 +1,1935 @@
...
+    out[12] = ((uint128_t) in1[6]) * in2[6];
+}
+
+/*-
+ * Reduce thirteen 128-bit coefficients to seven 64-bit coefficients.
+ * in[i] < 2^128 - 2^125
+ * out[i] < 2^56 for i < 6,
+ * out[6] <= 2^48
+ */
+static void felem_reduce(felem out, const widefelem in)
```
Comment (slontis) [inline crypto/ec/ecp_nistp384.c:451]: Is there any chance of explaining how you come up with all of the numbers?

Comment (rohanmclure) [reply crypto/ec/ecp_nistp384.c:451]: Are we talking about the inequalities or about the program constants?

When it comes to the subtraction constants, they are chosen as power of two multiples of $p_{384}$ that are sufficiently large so that each limb can be sufficiently large. This is achieved through a telescopic sum. You'll notice that we subtract $2^{68}$ from a high limb whenever we add $2^{124}$ to a lower limb. Since $2^{124} = 2^{56} \cdot 2^{68}$ these two alterations cancel out, but lead to each limb being roughly $2^{124}$ in size. This is helpful, as adding a multiple of our modulus like so allows us to avoid underflow when subtracting limbs, which is a requirement for correctness.

**Comment by rohanmclure** [comment]
Just updated `felem_reduce` to make `output[6]` larger and fix my bounds checks.

File: crypto/ec/ecp_nistp384.c [hunk 32]
```diff
@@ -0,0 +1,1931 @@
...
+# error "Your compiler doesn't appear to support 128-bit integer types"
+#endif
+
+typedef uint8_t u8;
+typedef uint64_t u64;
+
+/*
+ * The underlying field. P384 operates over GF(2^384-2^128-2^96+2^32-1). We
+ * can serialize an element of this field into 48 bytes where the most
+ * significant byte contains only a single bit. We call this an felem_bytearray.
```
Comment (slontis) [inline crypto/ec/ecp_nistp384.c:36]: Is the 'Most significant byte contains only a single bit' correct? (That would be 377 bits?)

Comment (rohanmclure) [reply crypto/ec/ecp_nistp384.c:36]: Apologies, yes this comment is completely wrong. I got this confused with `felem_reduce`, which returns a 385 bit number.

File: crypto/ec/ecp_nistp384.c [hunk 33]
```diff
@@ -0,0 +1,1931 @@
...
+ * felem_neg sets |out| to |-in|
+ * On entry:
+ *   in[i] < 2^60 - 2^29
+ * On exit:
+ *   out[i] < 2^60
+ */
+static void felem_neg(felem out, const felem in)
+{
+    /* In order to prevent underflow, we subtract from a multiple of p. */
+    static const limb two60m52m4 = (((limb) 1) << 60)
```
Comment (slontis) [inline crypto/ec/ecp_nistp384.c:215]: I am still struggling to get my head around how you come up with these static constants.. Can you give some insight into how you come up with the numbers below.. What is the actual multiple of p here? (x % p = 0 should be zero)..
I can see that the range seems top be 56 bits with the 2^60-2^4 part (with zeros in the top and bottom bits).

Comment (rohanmclure) [reply crypto/ec/ecp_nistp384.c:215]: In `felem_neg` it's just $2^{12} p_{384}$. See $2^{384} \cdot 2^{12} = 2^{60} \cdot (2^{56})^6$. We then use telescopic sums to introduce a $2^{60}$ term in every limb.

In each routine I use a simple power of two to get our multiple. Hope that makes sense.

Comment (rohanmclure) [reply crypto/ec/ecp_nistp384.c:215]: I may as well do this one in full:
$2^{12} p_{384} = 2^{396} - 2^{140} - 2^{108} + 2^{44} - 2^{12} = 2^{60} (2^{56})^6 - 2^{28} (2^{56})^2 - 2^{52} \cdot 2^{56} + 2^{44} - 2^{12}$
Now introducing into limbs 0 .. 5 a $2^60$ term:
$= (2^{60} - 2^4) (2^{56})^6 + (2^{60} - 2^4) (2^{56})^5 + (2^{60} - 2^4) (2^{56})^4 + (2^{60} - 2^4) (2^{56})^3 + (2^{60} - 2^{28} - 2^4) (2^{56})^2 + (2^{60} - 2^{52} - 2^4) 2^{56} + (2^{60} + 2^{44} - 2^{12})$

Comment (slontis) [reply crypto/ec/ecp_nistp384.c:215]: Thanks for explaining (and your patience).. That makes sense..
Is felem_diff_128_64 similiar?
Would you maybe be able to add comments explaining this in multiple functions, so others can also understand what is going on.

Comment (rohanmclure) [reply crypto/ec/ecp_nistp384.c:215]: Not a problem. Yep I was basically just copying the style of ecp_nistp{224,521}.c, but it doesn't hurt to just state the multiple up front.

Precisely, in `felem_diff_128_64` it's $2^{16}p_{384}$, and in `felem_reduce` it's $2^{76}p_{384}$.

**Comment by slontis** [comment]
I brought this PR up in the OTC meeting this week.
Hopefully @romen can also have a look at it. Most of it is similar to the other variants.
It would be easier to review if you were to explain in comments, how you came up with the values for the static constants, as you have done in the review conversation.
I wonder if there is some way to thrash test this, to make sure we don't have any 'bounds' problems. @romen mentioned there may have been issues previously with the other implementations.

File: crypto/ec/ecp_nistp384.c (hunk 33 above)
Comment (slontis) [reply crypto/ec/ecp_nistp384.c:215]: Is the felem_diff_128_64 still 2^56 based

**Comment by slontis** [comment]
Thank you for doing this PR. I wanted to add it, but didnt understand the maths in the other implementations..

I am wondering if others have opinions about if there should be multiple config options,
instead of just enable-ec_nistp_64_gcc_128?

Ideally p384 and p521 could be on by default (on systems that have support) since they are much faster...
The p256 case seems to be slightly slower.

(This doesnt have to be part of this PR)

**Comment by rohanmclure** [comment]
Yep good suggestion, I'll include these as comments.

I've written a suite of random tests with Python's `cffi` if that would help. Even still, it's best combined with the in-tree and external test suites, which include some important corner cases. Could something like this be useful?

**Comment by rohanmclure** [comment]
> Yep good suggestion, I'll include these as comments.
> 
> I've written a suite of random tests with Python's `cffi` if that would help. Even still, it's best combined with the in-tree and external test suites, which include some important corner cases. Could something like this be useful?



> I brought this PR up in the OTC meeting this week.
> Hopefully @romen can also have a look at it. Most of it is similar to the other variants.
> It would be easier to review if you were to explain in comments, how you came up with the values for the static constants, as you have done in the review conversation.

Yep good suggestion, I'll include these as comments.

> I wonder if there is some way to thrash test this, to make sure we don't have any 'bounds' problems. @romen mentioned there may have been issues previously with the other implementations.

I've written a suite of random tests with Python's cffi if that would help. Even still, it's best combined with the in-tree and external test suites, which include some important corner cases. Could something like this be useful?

**Comment by rohanmclure** [comment]
Updated license, unrolled loop for clarity in `felem_diff128`, and commented underflow constants.

File: crypto/ec/ecp_nistp384.c (hunk 33 above)
Comment (rohanmclure) [reply crypto/ec/ecp_nistp384.c:233]: Yes, so irrespective of the machine width - `widefelem` or `felem` and irrespective of the number ($k$) of limbs, the value being represented is always:

$\sigma^{k}_{i=0} (2^{56})^i l_i$

File: crypto/ec/ecp_nistp384.c [hunk 34]
```diff
@@ -0,0 +1,1975 @@
...
+    }
+    bin48_to_felem(out, b_out);
+    return 1;
+}
+
+/* felem_to_BN converts an felem into an OpenSSL BIGNUM */
+static BIGNUM *felem_to_BN(BIGNUM *out, const felem in)
+{
+    felem_bytearray b_out;
+    felem_to_bin48(b_out, in);
```
Comment (slontis) [inline crypto/ec/ecp_nistp384.c:163]: insert a blankline after the variable declarations

**Comment by slontis** [comment]
For anyone else trying to follow this the telescopic sum is given by:
$$\sum_{i=1}^n \left( 2^a - 2^b \right) \left( 2^\left( a-b \right)) \right)^i  = 2^\left(a+n(a-b)\right) - 2^a$$

File: crypto/ec/ecp_nistp384.c [hunk 35]
```diff
@@ -0,0 +1,1975 @@
...
+
+    output[0] += two124p108m76;
+    output[1] += two124m116m68;
+    output[2] += two124m92m68;
+    output[3] += two124m68;
+    output[4] += two124m68;
+    output[5] += two124m68;
+    output[6] += two124m68;
+
+    /* eliminate in[6], ..., in[12] */
```
Comment (slontis) [inline crypto/ec/ecp_nistp384.c:521]: Can you expand this comment to explain what the code below is doing.. I get the carry part but not this part..

Comment (rohanmclure) [reply crypto/ec/ecp_nistp384.c:521]: I'm deciding the most succinct way to comment this in code, but the underlying principle is this.
Let $p_{384} = 2^{384} - \delta$, then it follows that $2^{384} \equiv \delta \mod{p_{384}}$.

Thus for any suitable choice of $X$ and $Y$ such that our input is equal (or congruent) to $X + 2^{384}Y$ [1] which is in turn congruent to $X + \delta Y$.

The way you start off eliminating the high limbs (it should read 'eliminate in[7], ...., in[12]` is precisely with this substitution.

```c
    /* X += 2^128 Y */
    output[8] += in[12] >> 32;
    output[7] += (in[12] & 0xffffffff) << 24;

    /* X += 2^96 Y */
    output[7] += in[12] >> 8;
    output[6] += (in[12] & 0xff) << 48;
    
    /* X += (-2^32 + 1) Y */
    output[6] -= in[12] >> 16;
    output[5] -= ((in[12] & 0xffff) << 40);
    output[6] += in[12] >> 48;
    output[5] += (in[12] & 0xffffffffffff) << 8;
```

The reason why at each limb, and for term of $\delta = 2^{128} + 2^{96} + (-2^{32} + 1)$ there are multiple operations being performed is simply to reduce the maximum size of operands to prevent under/overflow.

in reality we do at least three such substitutions in this code. Firstly we use limbs `in[12]`, `in[11]`, `in[10]`, `in[9]` to be our $Y$. Then afterwards we take `output[8]`, `output[7]` as our $Y$ (the result of the previous substitution accumulated into these), eliminating these two limbs. Finally, we let `temp` be our choice of $Y$ (each of these with suitable scaling so that [1] holds) and substitute it (after partially carrying into `temp` in order to minimise the size of the resulting, reduced number). We perform one last final carry and then obtain the desired bound.

File: crypto/ec/ecp_nistp384.c [hunk 36]
```diff
@@ -0,0 +1,1975 @@
+/*
+ * Copyright 2023 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/* Copyright 2023 IBM Corp.
```
Comment (slontis) [inline crypto/ec/ecp_nistp384.c:10]: /*
 * Copyright

File: crypto/ec/ecp_nistp384.c [hunk 37]
```diff
@@ -0,0 +1,1975 @@
...
+
+    /*
+     * In effect, invert our conditions, so that 0 values become all 1's,
+     * any non-zero value in the low-order 56 bits becomes all 0's
+     */
+    for (i = 0; i < 5; i++)
+       cond[i] = ((cond[i] & 0x00ffffffffffffff) - 1) >> 63;
+
+    /*
+     * The condition for determining whether in is greather than our
```
Comment (slontis) [inline crypto/ec/ecp_nistp384.c:903]: greater

File: crypto/ec/ecp_nistp384.c (hunk 36 above)
Comment (paulidale) [reply crypto/ec/ecp_nistp384.c:10]: I don't see why not.  There are plenty of other files with dual copyrights.

**Comment by slontis** [comment]
Apart from some minor NITS this looks good to me.

**Comment by rohanmclure** [comment]
Thanks for your comments. I've updated `felem_reduce`, giving it a header-comment providing an overview of the elimination stages, extended the in-line comments and replaced the `output` variable with `acc`.

**Review by paulidale — APPROVED** [review]
Nice work and thanks for persevering.

**Review by slontis — APPROVED** [review]

**Comment by tmshort** [comment]
I'd merge this, but @beldmit has an outstanding "change requested".

**Comment by tmshort** [comment]
Merged to `master`, thank you for your contribution!
